File: c++/maximum-depth-of-binary-tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode *root) {
        /* https://oj.leetcode.com/problems/maximum-depth-of-binary-tree/
        iteratively, postorder traversal
        */
        
        stack<TreeNode*> st;
        TreeNode *p = root;
        int depth = 0, result = 0;
        
        while (p != NULL) {
            depth++;
            st.push(p);
            if (p->left != NULL) {
                p = p->left;
            }
            else if (p->right != NULL) {
                p = p->right;
            }
            else { // p is a leaf node
                if (depth > result) result = depth;
                depth--;
                st.pop();
                
                while (!st.empty()) {
                    if (st.top()->right == p || st.top()->right == NULL) {
                        // it's time to visit st.top()
                        p = st.top();
                        depth--;
                        st.pop();
                    }
                    else {
                        p = st.top()->right;
                        break;
                    }
                }
                
                if (st.empty()) break;
            }
        }
        
        return result;
    }
};

File: python/maximum-depth-of-binary-tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return an integer
    def maxDepth(self, root):
        if not root:
            return 0

        max_depth = 0
        cur = root
        depth = 0
        stack = []

        while cur:
            stack.append(cur)
            depth += 1
            if cur.left:
                cur = cur.left
            elif cur.right:
                cur = cur.right
            else:
                # meet a leaf node
                if depth > max_depth:
                    max_depth = depth

                while stack:
                    if stack[-1].right == cur or stack[-1].right is None:
                        cur = stack.pop()
                        depth -= 1
                    else:
                        cur = stack[-1].right
                        break

                if not stack:
                    break

        return max_depth
